### DISCUSSION TITLE
Exploiting the -timestamp capability

### DISCUSSION THREAD

**Discussion by ewillink** [eclipse-justj/justj.tools#7] [Q&A] [answered]
JustJ allows a timestamp to be specified, but not in a way that is readily compatible.

Traditionally the time stamp is set to yyyyMMdd-HHmm for plugins and features and yyyyMMddHHmm for builds. JustJ enforces the latter and rejects an attempt to pass the former for re-use. There do not appear to be any readily available string editing functionalities between the Tycho session that builds and a subsequent Tycho session that uses JustJ. (The functionality that JustJ is displacing ran at shell level so shell string maniuplation was available to strip the "-").

Is there a simple solution?

Workaround omit the -timestamp and it defaults to 'now' that is a few minutes late. and different to the contents.

**Comment by merks** [comment, marked as answer]
You can set this property in the pom.xml where you do the promotion step to determine the format of the `${build.timestamp}` substitution in that pom..
```
  <properties>
    <maven.build.timestamp.format>yyyyMMddHHmm</maven.build.timestamp.format>
  </properties>
```

**Reply by ewillink** [reply]
This helps a bit. The features and plugins still use a "-" date based on the start of the build, and now the exported timestamp uses a non-"-" date but it is close to the end of the build. Maybe the build-site pom needs something in an early phase to force early resolution of the maven.build.timestamp.format property. (I struggle to find coherent listings of the phases once enhanced by Tycho and extra plugins.)

**Reply by merks** [reply]
You're making assumptions.  You assume that the value changes during the lifetime of build and you assume the expansion of `${...}` happens later than is actually the case.  The value itself is the time when the build is started, though it can be formatted differently in different poms.  Even if that weren't the case, the expansion happens very early when reading all the poms and interpreting what they specify; so the replacement is fixed at the start of the build not dynamic during the build.

**Reply by ewillink** [reply]
oops. Sorry. I passed the timestamp without a -timestamp prefix, so of course it used 'now' but surprisingly didn't fail the command line. Works great now.
